### PR TITLE
Reduce stencil call overhead

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Purpose

The current `fv3core` contains two stencil wrapper classes, `FV3StencilObject` and `FixedOriginStencil`. Both call the `gtscript.StencilObject._call_run` method that contains a considerable amount of overhead necessary to generate the information needed to call the stencil and to validate the input arguments. This PR introduces a single class, `StencilWrapper`, that encapsulates some of the logic in both existing wrappers, but calls the the `gtscript.StencilObject.run` method directly to reduce stencil call overhead time.

## Code changes:

- Combine `StencilWrapper` and `FrozenStencil` classes.
- Make `FV3StencilObject` a subclass of `StencilWrapper`.
- Call `stencil_object.run` methods directly.
- Replace `ValueError` exceptions at stencil call time with assertions so `PYTHONOPTIMIZED` will omit them.
- Replace all occurrences of `FrozenStencil` with `StencilWrapper` in the OOP-ified stencil classes.
- Cache the results of the expensive `_compute_field_origins` method in the wrapper object.
- Add `StencilObjectCache` to hold `StencilWrapper` objects and enable clearing of cached information between tests.

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate
- [x] All relevant documentation has been updated or added (e.g. README, CONTRIBUTING docs)

### CODE REVIEW

**Comment by eddie-c-davis** [comment]
launch jenkins

File: fv3core/decorators.py [hunk 1]
```diff
@@ -149,7 +149,96 @@ def __iter__(self):
...
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+            **kwargs,
+        )
+
+    def __call__(self, *args, **kwargs) -> None:
+        if self.origin:
+            assert "origin" not in kwargs, "Cannot override origin provided at init"
+            kwargs["origin"] = self.origin
+        kwargs["origin"] = self._get_field_origins(*args, **kwargs)
```
Comment (twicki) [inline fv3core/decorators.py:180]: I think we should store this information to only compute it once. 
```suggestion
        if not self.fields_with_origin:
           self.fields_with_origin = self._get_field_origins(*args, **kwargs)
[...]
elf.stencil_object.run(**kwargs, origin=self.fields_with_origin),
```

File: fv3core/decorators.py [hunk 2]
```diff
@@ -149,7 +149,96 @@ def __iter__(self):
...
+        if global_config.get_validate_args():
+            self.stencil_object(*args, **kwargs, validate_args=True)
+        else:
+            kwargs = self._process_kwargs(*args, **kwargs)
+            self.stencil_object.run(**kwargs, exec_info=None)
+
+    def _process_kwargs(self, *args, **kwargs):
+        """Process keyword args so they can be passed directly to stencil_object.run."""
+
+        for keyword in ("origin", "domain"):
```
Comment (twicki) [inline fv3core/decorators.py:197]: pop with None as the secondary value lets you delete if present if that is easier?

File: fv3core/decorators.py [hunk 3]
```diff
@@ -149,7 +149,96 @@ def __iter__(self):
...
+        kwargs["origin"] = self._get_field_origins(*args, **kwargs)
+
+        if self.domain:
+            assert "domain" not in kwargs, "Cannot override domain provided at init"
+            kwargs["domain"] = self.domain
+        else:
+            assert "domain" in kwargs, "No domain provided at call time"
+
+        if global_config.get_validate_args():
+            self.stencil_object(*args, **kwargs, validate_args=True)
```
Comment (twicki) [inline fv3core/decorators.py:189]: why is exec info not passed here?

File: fv3core/decorators.py [hunk 4]
```diff
@@ -149,7 +149,96 @@ def __iter__(self):
...
+            kwargs["origin"] = self.origin
+        kwargs["origin"] = self._get_field_origins(*args, **kwargs)
+
+        if self.domain:
+            assert "domain" not in kwargs, "Cannot override domain provided at init"
+            kwargs["domain"] = self.domain
+        else:
+            assert "domain" in kwargs, "No domain provided at call time"
+
+        if global_config.get_validate_args():
```
Comment (twicki) [inline fv3core/decorators.py:188]: should we not just `validate_args= global_config.get_validate_args()`?

File: fv3core/decorators.py [hunk 5]
```diff
@@ -149,7 +149,96 @@ def __iter__(self):
...
 
 
-class FV3StencilObject:
+class StencilWrapper:
+    """Wrapped GT4Py stencil object."""
+
+    def __init__(
+        self,
+        func: Callable[..., None],
+        origin: Optional[Index3D] = None,
```
Comment (twicki) [inline fv3core/decorators.py:158]: with this implementation we don't really have a benefit any more for initializing with origin and domain. Do we still want to do it nevertheless for readability or should we revert that functionality?

**Review by twicki — CHANGES_REQUESTED** [review]

File: fv3core/decorators.py (hunk 4 above)
Comment (eddie-c-davis) [reply fv3core/decorators.py:188]: Changed to an actual keyword argument per Jeremy's suggestion.

File: fv3core/decorators.py (hunk 3 above)
Comment (eddie-c-davis) [reply fv3core/decorators.py:189]: Nice catch!

File: fv3core/decorators.py (hunk 2 above)
Comment (eddie-c-davis) [reply fv3core/decorators.py:197]: Good idea!

File: fv3core/decorators.py (hunk 1 above)
Comment (eddie-c-davis) [reply fv3core/decorators.py:180]: Implemented, testing in progress...

File: fv3core/decorators.py (hunk 5 above)
Comment (eddie-c-davis) [reply fv3core/decorators.py:158]: I feel like it is good to keep it for explicitness.

File: fv3core/decorators.py [hunk 6]
```diff
@@ -150,16 +150,169 @@ def __iter__(self):
...
+        return field_origins
+
+    def _set_device_sync(self, stencil_kwargs: Dict[str, Any]):
+        """Sets the 'device_sync' backend option in the stencil kwargs."""
+        if self.device_sync is None and "cuda" in self.backend:
+            self.device_sync = global_config.get_device_sync()
+        if self.device_sync is not None:
+            stencil_kwargs["device_sync"] = self.device_sync
+
+    @property
```
Comment (twicki) [inline fv3core/decorators.py:299]: if we make these two properties, should we also create proper setters?

**Review by twicki — CHANGES_REQUESTED** [review]

File: fv3core/decorators.py [hunk 7]
```diff
@@ -150,16 +150,165 @@ def __iter__(self):
...
-        self.stencil_object: Optional[gt4py.StencilObject] = None
+        self.origin: Tuple[int, ...] = origin
+        """The compute origin."""
+
+        if domain is not None:
+            domain = Shape(domain)
+        self.domain: Optional[Shape] = domain
+        """The compute domain."""
+
+        self.backend: str = backend if backend else global_config.get_backend()
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply fv3core/decorators.py:184]: Modified to only support the `global_config` versions.

File: fv3core/decorators.py [hunk 8]
```diff
@@ -150,16 +150,169 @@ def __iter__(self):
...
 
-    def __init__(self, func: Callable[..., None], **kwargs):
-        self.func: Callable[..., None] = func
+    def __init__(
+        self,
+        func: Callable[..., None],
+        origin: Optional[Tuple[int, ...]] = None,
+        domain: Optional[Index3D] = None,
+        *,
+        disable_cache: bool = True,
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply fv3core/decorators.py:162]: Good point, changed.

File: fv3core/decorators.py [hunk 9]
```diff
@@ -150,16 +150,169 @@ def __iter__(self):
...
+    ) -> None:
+        if self.origin:
+            assert origin is None, "cannot override origin provided at init"
+            origin = self.origin
+        if not self.field_origins or self.disable_cache:
+            self.field_origins = self._compute_field_origins(origin, *args, **kwargs)
+
+        if self.domain:
+            assert domain is None, "cannot override domain provided at init"
+            domain = self.domain
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply fv3core/decorators.py:237]: So it can be passed to `self.stencil_object` if `validate_args` is `True`.

File: fv3core/decorators.py (hunk 6 above)
Comment (eddie-c-davis) [reply fv3core/decorators.py:299]: These are read-only convenience functions to extract information from the `stencil_object`, so I would say no setters.

File: fv3core/decorators.py [hunk 10]
```diff
@@ -150,16 +150,164 @@ def __iter__(self):
...
+
+    def _process_kwargs(self, domain: Optional[Index3D], *args, **kwargs):
+        """Processes keyword args for direct calls to stencil_object.run."""
+
+        kwargs["_origin_"] = self.field_origins
+        kwargs["_domain_"] = domain
+
+        arg_names = self.field_names + self.parameter_names
+        for i in range(len(args)):
+            kwargs[arg_names[i]] = args[i]
```
Comment (jdahm) [inline fv3core/decorators.py:259]:
```suggestion
        arg_names = self.field_names + self.parameter_names
        kwargs.update({name: arg for name, arg in zip(arg_names, args)})
```

File: fv3core/decorators.py [hunk 11]
```diff
@@ -150,16 +150,164 @@ def __iter__(self):
...
+        return kwargs
+
+    def _compute_field_origins(
+        self, origin: Tuple[int, ...], *args, **kwargs
+    ) -> Dict[str, Tuple[int, ...]]:
+        """Computes the origin for each field in the stencil call."""
+
+        field_origins: Dict[str, Tuple[int, ...]] = {"_all_": origin}
+        field_names: List[str] = self.field_names
+        for i in range(len(field_names)):
```
Comment (jdahm) [inline fv3core/decorators.py:270]:
```suggestion
        for arg, name in zip(args, field_names):
```

File: fv3core/decorators.py [hunk 12]
```diff
@@ -150,16 +150,169 @@ def __iter__(self):
...
+        if validate_args:
+            self.stencil_object(
+                *args,
+                **kwargs,
+                origin=self.field_origins,
+                domain=domain,
+                validate_args=True,
+            )
+        else:
+            kwargs = self._process_kwargs(domain, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply fv3core/decorators.py:248]: Implemented...

File: fv3core/decorators.py [hunk 13]
```diff
@@ -150,16 +150,169 @@ def __iter__(self):
...
+        """The gt4py generated code is formatted if true."""
+
+        self.device_sync: bool = device_sync
+        """Synchronize device (GPU) after each stencil call if true."""
+
+        self.field_origins: Dict[str, Tuple[int, ...]] = {}
+        """A dictionary of data field origins."""
+
+        stencil_object = None
+        if not delay_compile:
```
[the remark this answers is not in the document]
Comment (twicki) [reply fv3core/decorators.py:203]: great

**Review by twicki — APPROVED** [review]
great, thanks a lot!

**Pull request merged**